esp-hal/src/dma/mod.rs

```diff
@@ -1852,6 +1872,16 @@ pub trait DmaTxBuffer {
     fn length(&self) -> usize;
 }
 
+impl<T: DmaTxBuffer> DmaTxBuffer for &mut T {
```

Comment:
Oh no I was going to explicitly block this in an upcoming PR, this violates the safety guarantees. I'm guessing this is what you were talking about in my PR?

Reply:
The proper way to do this would be to use a wrapper around it with an unsafe constructor.

Reply:
I tried it in one of the dead-ends, but this was the least amount of code in an already big PR. These buffers will be unsound in `transfer_in_place` because there they'll have to point to the same data and they currently use mutable reference for that, potentially violating aliasing rules.

I have a refactor stashed somewhere that extracts a RawBuffer that stores ptr and len separately in the struct.

Reply:
I don't think this violates anything. The transfer object owns a mutable reference, so nobody is allowed to touch the buffer while it's in flight. What am I missing, besides days' worth of sleep?

Reply:
Leak-apocalypse. My least favorite thing about Rust. `core::mem::forget` can be called on the transfer object.

Reply:
Sorry, I went with a "let me jug your memory" type of answer. https://docs.rust-embedded.org/embedonomicon/dma.html#memforget describes the safety problem, just in case you haven't seen it before.

Reply:
![image](https://github.com/user-attachments/assets/4bbc4a3a-d1c5-41ef-9f4b-52990daa050c)

> My least favorite thing about Rust.

Well, we agree here. But the project seems to have chosen to ignore the problem, as this is largely just "I want to commit crimes against my hardware" kind of situation.

Reply:
Yeah that was added in for the DMA in #1245 and I was working towards undoing that with the DMA bufs. The SPI driver is currently `forget`-safe and I'd rather not undo it unless it gets in the way of adding something. Zerocopy can still be added without sacrificing the safety, or at least any sacrifices in safety can be made in a layer above the safe layer, with a custom `DmaBuffer` implementation.

Reply:
Alright let me try and clean this up

Reply:
I keep thinking that this disclaimer is something we need to keep. I think I see what you mean by "Zerocopy can still be added without sacrificing the safety", but I don't see if we can apply that to the async API, at least not the embedded-hal implementation. Taking a slice there, then forgetting the transfer will always be A Bad Thing (TM). And I really, really want that API to not copy the data if it doesn't need to.

I think regardless of whether we can come up with a forget-safe API, the user will rarely want to forget anything we return to them. In case of transfer objects, they'd lose access to the peripheral. In case of an async operation, well, what would the user expect, anyway?

Reply:
Yes it doesn't apply in the async case since the user can choose to forget the future. Arguably copying is the "right thing to do" for async e-hal APIs (ideally with some double buffering so the impact of copying is minimal), especially because it lets you return early, but I'm not going to push this aspect.

(Though now that we've brought up e-hal, I'm now wondering if cancelling the transfer is even the correct thing to do based on the docs of `SpiBus`, the transfer should continue in the background until `flush()` is called?)

My main point here, really, is not about making the entire driver forget-safe. It's about keeping the low-level/advanced APIs safe, and building the unsafe/unsound stuff on top of it. "The foundation should be solid".
Sure, this means the disclaimer stays, unless we mark the `SpiDmaBus` constructor as unsafe but I don't think people would like that very much.

Reply:
> (Though now that we've brought up e-hal, I'm now wondering if cancelling the transfer is even the correct thing to do based on the docs of SpiBus, the transfer should continue in the background until flush() is called?)

Dang that is a very good point...

Reply:
Though the embassy-stm32 impl is also "yeah lol whatever" around these questions, it seems.

Reply:
> Though the embassy-stm32 impl is also "yeah lol whatever" around these questions, it seems.

Bahaha, no one can be bothered it seems. Hopefully we can have robust answers in esp-hal.